pkg/compiler/validators/utils.go

```diff
@@ -153,43 +153,72 @@ func UnionDistinctVariableMaps(m1, m2 map[string]*core.Variable) (map[string]*co
 	return res, nil
 }
 
-// Gets LiteralType for literal, nil if the value of literal is unknown, or type None if the literal is a non-homogeneous
-// type.
+// LiteralTypeForLiteral gets LiteralType for literal, nil if the value of literal is unknown, or type collection/map of
+// type None if the literal is a non-homogeneous type.
 func LiteralTypeForLiteral(l *core.Literal) *core.LiteralType {
 	switch l.GetValue().(type) {
 	case *core.Literal_Scalar:
 		return literalTypeForScalar(l.GetScalar())
 	case *core.Literal_Collection:
 		if len(l.GetCollection().Literals) == 0 {
-			return &core.LiteralType{Type: &core.LiteralType_Simple{Simple: core.SimpleType_NONE}}
+			return &core.LiteralType{
+				Type: &core.LiteralType_CollectionType{
+					CollectionType: &core.LiteralType{
+						Type: &core.LiteralType_Simple{Simple: core.SimpleType_NONE},
+					},
+				},
+			}
 		}
 
 		// Ensure literal collection types are homogeneous.
 		var innerType *core.LiteralType
 		for _, x := range l.GetCollection().Literals {
 			otherType := LiteralTypeForLiteral(x)
+
 			if innerType != nil && !AreTypesCastable(otherType, innerType) {
-				return &core.LiteralType{Type: &core.LiteralType_Simple{Simple: core.SimpleType_NONE}}
+				return &core.LiteralType{
+					Type: &core.LiteralType_CollectionType{
+						CollectionType: &core.LiteralType{
+							Type: &core.LiteralType_Simple{Simple: core.SimpleType_NONE},
+						},
+					},
+				}
```

Comment:
I'm not sure I fully understand this. In this case if we find that the collection is not homogenous (ex. [9.99, "foo"]) then return [None], but this change explicitly allows for [None] to be castable to any type (ex. [Int], [String], etc). Is there a potential bug here where [9.99, "foo"] is castable to [Int]? Can we add a unit test for this case?

Reply:
so originally I left it to that yes because Array types support this, very awkward for various reasons, feature where you can specify minimum successes (or success ratio). When some tasks fail in an array task, their results will be substituted with `None` in the final assembled array output. This means a consuming task may receive `[10, None, 1, 3, 4]` even when it doesn't say `List[Optional[int]]` as the type... so I thought I would leave that behavior in..

However, now that you asked the question, I went back and re-read through and we don't actually go through this path for array tasks... so this is ok... 

I updated the code to, I think more correctly, build a `List[Union[int, string]]` in case the input is `[10, "foo"]` or `List[Union[int, None]]` (essentially `List[Optional[int]]` in case of `[10, None]`... wdyt?

Reply:
Perfect, I think this makes more sense!